pydatastructs/trees/binary_trees.py

```diff
@@ -1057,6 +1047,357 @@ def split(self, x):
+            self.tree[self.root_idx].data = None
+            self.tree[self.root_idx].key = None
+        elif self._has_one_child(node_idx):
+            root_key = self._transplant_values(node_idx, node_idx1)
+            new_indices = self.tree.delete(node_idx1)
+            if new_indices is not None:
+                self.root_idx = new_indices[root_key]
+        return
+
+    def __leaf_case(self, node_idx, node_idx1):
+        walk = node_idx
+        walk1 = node_idx1
+        parent = self._get_parent(node_idx)
+        color = self.tree[walk].color
+        if parent is None:
+            self._delete_root(walk, walk1)
+        else:
+            if self.__walk1_walk_isblack(color, walk1):
+                self.__fix_deletion(walk)
+            else:
+                sibling_idx = self._get_sibling(walk)
+                if sibling_idx is not None:
+                    self.tree[sibling_idx].color = 1
+            if self._is_onleft(walk):
+                self.tree[parent].left = None
+            else:
+                self.tree[parent].right = None
+            self._remove_node(walk)
+
+    def __one_child_case(self, node_idx, node_idx1):
+        walk = node_idx
+        walk1 = node_idx1
+        walk_original_color = self.tree[walk].color
+        parent = self._get_parent(node_idx)
+        if parent is None:
+            self._delete_root(walk, walk1)
+        else:
+            if self._is_onleft(walk):
+                self.tree[parent].left = walk1
+            else:
+                self.tree[parent].right = walk1
+            self.tree[walk1].parent = parent
+            a = self._remove_node(walk)
+            if self.__walk1_walk_isblack(walk_original_color, walk1):
+                self.__fix_deletion(walk1)
+            else:
+                self.tree[walk1].color = 0
+
+    def __two_child_case(self, node_idx):
+        walk = node_idx
+        successor = self._replace_node(walk)
+        self._transplant_values(walk, successor)
+        walk = successor
+        walk1 = self._replace_node(walk)
+        return walk, walk1
+
+    def delete(self, key, **kwargs):
+        walk = super(Redblacktree, self).search(key)
+        if walk is not None:
+            walk1 = self._replace_node(walk)
+            if self._has_two_child(walk):
+                walk, walk1 = self.__two_child_case(walk)
+            if self._is_leaf(walk):
+                self.__leaf_case(walk, walk1)
+            elif self._has_one_child(walk):
+                self.__one_child_case(walk, walk1)
+            return True
+        else:
+            return None
+
+    def __str__(self):
+        to_be_printed = ['' for i in range(self.tree._last_pos_filled + 1)]
+        for i in range(self.tree._last_pos_filled + 1):
+            if self.tree[i] is not None:
+                node = self.tree[i]
+                to_be_printed[i] = (node.left, node.key, node.color, node.data, node.right)
+        return str(to_be_printed)
```

Comment:
I think, `node.color` isn't a concern of the end user. You can remove this method.